scripts/sensor/03b-time_frequency_csp.py

```diff
@@ -0,0 +1,866 @@
+
+        epochs_filter, y = prepare_epochs_and_y(
+            epochs=epochs, fmin=fmin, fmax=fmax, cfg=cfg)
+
+        X = epochs_filter.get_data()
+
+        # Save mean scores over folds
+        cv_scores = cross_val_score(estimator=clf, X=X, y=y,
+                                    scoring='roc_auc', cv=cv,
+                                    n_jobs=1)
+
+        freq_scores_std[freq] = np.std(cv_scores, axis=0)
+        freq_scores[freq] = np.mean(cv_scores, axis=0)
+
+    np.save(file=pth.freq_scores(subject), arr=freq_scores)
+    np.save(file=pth.freq_scores_std(subject), arr=freq_scores_std)
+    fig = plot_frequency_decoding(
+        freqs=tf.freqs,
+        freq_scores=freq_scores,
+        conf_int=freq_scores_std,
+        pth=pth,
+        subject=subject)
+
+    report.add_figs_to_section(  # TODO: That does not work
+        fig,
+        section="Frequency roc-auc decoding",
+        captions=f'sub-{subject}')
+
+    ######################################################################
+    # 2. Loop through frequencies and time
+
+    # init scores
+    tf_scores = np.zeros((tf.n_freq_windows, tf.n_time_windows))
+
+    # Loop through each frequency range of interest
+    for freq, (fmin, fmax) in tqdm(
+            enumerate(tf.freq_ranges),
+            total=tf.n_freq_windows,
+            desc=f'subject {subject} - time-frequency loop'):
+
+        epochs_filter, y = prepare_epochs_and_y(
+            epochs=epochs, fmin=fmin, fmax=fmax, cfg=cfg)
+
+        # Roll covariance, csp and lda over time
+        for t, (w_tmin, w_tmax) in enumerate(tf.time_ranges):
+
+            # Originally the window size varied accross frequencies...
+            # But this means also that there is some mutual information between
+            # 2 different pixels in the map.
+            # So the simple way to deal with this is just to fix
+            # the windows size for all frequencies.
+
+            # Crop data into time-window of interest
+            X = epochs_filter.copy().crop(w_tmin, w_tmax).get_data()
+
+            # Save mean scores over folds for each frequency and time window
+            cv_scores = cross_val_score(estimator=clf,
+                                        X=X, y=y,
+                                        scoring='roc_auc',
+                                        cv=cv,
+                                        n_jobs=1)
+
+            # TODO: add CSP plots
+
+    np.save(file=pth.tf_scores(subject), arr=tf_scores)
+    fig = plot_time_frequency_decoding(
+        freqs=tf.freqs, tf_scores=tf_scores, sfreq=sfreq, pth=pth,
+        centered_w_times=tf.centered_w_times, subject=subject)
+    report.add_figs_to_section(
+        fig,
+        section="Time - frequency decoding",
+        captions=f'sub-{subject}')
+
+    msg = f"Decoding for subject {subject} finished successfully."
+    logger.info(gen_log_message(message=msg, subject=subject, step=3))
+
+
+def load_and_average(
```

Comment:
why do you need this?

Reply:
I agree it's not necessary to dump and then load the numpy arrays containing the tf_scores and frequency_scores, but since the function one_subject_decoding takes a long time to compute, and since the group analysis function is very cheap computationally, it's easy to just comment the lines where I launch the one_subject_decoding in order to just recalculate the group statistics by changing the parameters like the tstat value to rapidly iterate with the plotting. Also, At the beginning, it was not clear if we would split the script into 2 scripts between the csp estimation and the permutation test.